examples/vits.py

```diff
@@ -0,0 +1,725 @@
+import json, logging, math, os, re, sys, time, wave, argparse, numpy as np, inflect
+from pathlib import Path
+from typing import List
+from phonemizer import phonemize
+from unidecode import unidecode
+from extra.utils import download_file
+from tinygrad import nn
+from tinygrad.helpers import dtypes
+from tinygrad.state import torch_load
+from tinygrad.tensor import Tensor
+
+LRELU_SLOPE = 0.1
+
+class Synthesizer:
+  def __init__(self, n_vocab, spec_channels, segment_size, inter_channels, hidden_channels, filter_channels, n_heads, n_layers, kernel_size, p_dropout, resblock, resblock_kernel_sizes, resblock_dilation_sizes, upsample_rates, upsample_initial_channel, upsample_kernel_sizes, n_speakers=0, gin_channels=0, use_sdp=True, emotion_embedding=False, **kwargs):
+    self.n_vocab, self.spec_channels, self.inter_channels, self.hidden_channels, self.filter_channels, self.n_heads, self.n_layers, self.kernel_size, self.p_dropout, self.resblock, self.resblock_kernel_sizes, self.resblock_dilation_sizes, self.upsample_rates, self.upsample_initial_channel, self.upsample_kernel_sizes, self.segment_size, self.n_speakers, self.gin_channels, self.use_sdp = n_vocab, spec_channels, inter_channels, hidden_channels, filter_channels, n_heads, n_layers, kernel_size, p_dropout, resblock, resblock_kernel_sizes, resblock_dilation_sizes, upsample_rates, upsample_initial_channel, upsample_kernel_sizes, segment_size, n_speakers, gin_channels, use_sdp
+    self.enc_p = TextEncoder(n_vocab, inter_channels, hidden_channels, filter_channels, n_heads, n_layers, kernel_size, p_dropout, emotion_embedding)
+    self.dec = Generator(inter_channels, resblock, resblock_kernel_sizes, resblock_dilation_sizes, upsample_rates, upsample_initial_channel, upsample_kernel_sizes, gin_channels=gin_channels)
+    self.enc_q = PosteriorEncoder(spec_channels, inter_channels, hidden_channels, 5, 1, 16, gin_channels=gin_channels)
+    self.flow = ResidualCouplingBlock(inter_channels, hidden_channels, 5, 1, 4, gin_channels=gin_channels)
+    self.dp = StochasticDurationPredictor(hidden_channels, 192, 3, 0.5, 4, gin_channels=gin_channels) if use_sdp else DurationPredictor(hidden_channels, 256, 3, 0.5, gin_channels=gin_channels)
+    if n_speakers > 1: self.emb_g = nn.Embedding(n_speakers, gin_channels)
+  def infer(self, x, x_lengths, sid=None, noise_scale=1.0, length_scale=1, noise_scale_w=1., max_len=None, emotion_embedding=None):
+    x, m_p, logs_p, x_mask = self.enc_p.forward(x, x_lengths, emotion_embedding)
+    g = self.emb_g(sid.reshape(1, 1)).squeeze(1).unsqueeze(-1) if self.n_speakers > 0 else None
+    logw = self.dp.forward(x, x_mask, g=g, reverse=self.use_sdp, noise_scale=noise_scale_w if self.use_sdp else 1.0)
+    w_ceil = Tensor.ceil(logw.exp() * x_mask * length_scale)
+    y_lengths = Tensor.maximum(w_ceil.sum([1, 2]), 1).cast(dtypes.int64)
+    y_mask = sequence_mask(y_lengths, y_lengths.max().lazydata.toCPU()).unsqueeze(1).cast(x_mask.dtype)
```

Comment:
We should find a better way to do this.

Reply:
So I added a very hacky way to estimate the output sequence length, which avoids realisation up until the end.

Once the model reaches the end, I realise the output, then realise the `y_lengths` buffer, then compare the actual `y_length` to the estimate, and if the estimated length was too short, it will recompute without the estimate, if the estimate was too long it will trim it. 
```py
if max_y_length_estimate_scale is not None:
  length_scaler = o.shape[-1] / max_y_length
  o.realize()
  real_max_y_length = y_lengths.max().numpy()
  if real_max_y_length > max_y_length:
    logging.warning(f"Underestimated max length by {(((real_max_y_length / max_y_length) * 100) - 100):.2f}%, recomputing inference without estimate...")
    return self.generate(g, logs_p, m_p, max_len, None, noise_scale, w_ceil, x, x_mask, y_lengths)
  if real_max_y_length < max_y_length:
    overestimation = ((max_y_length / real_max_y_length) * 100) - 100
    logging.info(f"Overestimated max length by {overestimation:.2f}%")
    if overestimation > 10: logging.warning("Warning: max length overestimated by more than 10%")
  o = o[:, :, :(real_max_y_length * length_scaler).astype(np.int32)]
return o
```
I also added a warning in case the overestimation is too big, this indicates to the user that maybe for their use case they need to modify the `max_y_length_estimate_scale` or provide none entirely.

Reply:
This is better than the lazydata usage, but it is slower for me than not doing a length estimation by ~0.05 seconds.

Reply:
True, for smaller inputs its actually a tad slower, for larger inputs it should be a lot faster tho.

Reply:
Also `Y_LENGTH_ESTIMATE_SCALARS` for larger inputs can be set lower, there seems to be a logarithmic relation between input size and the length, but for smaller inputs it varies a lot depending on the speaker and symbols.  For larger inputs it seems to converge to about `1.56` for the vctk model.